wallet/recovery_test.go

```diff
@@ -0,0 +1,245 @@
+		total uint32
+	}
+
+	// MarkInvalid is a Step that marks the `child` as invalid in the branch
+	// recovery state.
+	MarkInvalid struct {
+		child uint32
+	}
+
+	// ReportFound is a Step that reports `child` as being found to the
+	// branch recovery state.
+	ReportFound struct {
+		child uint32
+	}
+)
+
+// Apply extends the current horizon of the branch recovery state, and checks
+// that the returned delta is equal to the test's recovery window. If the
+// assertions pass, the harness's expected horizon is increased by the returned
+// delta.
+//
+// NOTE: This should be used before applying any CheckDelta steps.
+func (_ InitialDelta) Apply(i int, h *Harness) {
+	curHorizon, delta := h.brs.ExtendHorizon()
+	assertHorizon(h.t, i, curHorizon, h.expHorizon)
+	assertDelta(h.t, i, delta, h.recoveryWindow)
+	h.expHorizon += delta
+}
+
+// Apply extends the current horizon of the branch recovery state, and checks
+// that the returned delta is equal to the CheckDelta's child value.
+func (d CheckDelta) Apply(i int, h *Harness) {
+	curHorizon, delta := h.brs.ExtendHorizon()
+	assertHorizon(h.t, i, curHorizon, h.expHorizon)
+	assertDelta(h.t, i, delta, d.delta)
+	h.expHorizon += delta
+}
+
+// Apply queries the branch recovery state for the number of invalid children
+// that lie between the last found address and the current horizon, and compares
+// that to the CheckNumInvalid's total.
+func (m CheckNumInvalid) Apply(i int, h *Harness) {
+	assertNumInvalid(h.t, i, h.brs.NumInvalidInHorizon(), m.total)
+}
+
+// Apply marks the MarkInvalid's child index as invalid in the branch recovery
+// state, and increments the harness's expected horizon.
+func (m MarkInvalid) Apply(i int, h *Harness) {
+	h.brs.MarkInvalidChild(m.child)
+	h.expHorizon++
+}
+
+// Apply reports the ReportFound's child index as found in the branch recovery
+// state. If the child index meets or exceeds our expected next unfound value,
+// the expected value will be modified to be the child index + 1. Afterwards,
+// this step asserts that the branch recovery state's reportednext unfound value
+// matches our potentially-updated value.
+func (r ReportFound) Apply(i int, h *Harness) {
+	h.brs.ReportFound(r.child)
+	if r.child >= h.expNextUnfound {
+		h.expNextUnfound = r.child + 1
+	}
+	assertNextUnfound(h.t, i, h.brs.NextUnfound(), h.expNextUnfound)
+}
+
+// Compile-time checks to ensure our steps implement the Step interface.
+var _ Step = InitialDelta{}
+var _ Step = CheckDelta{}
+var _ Step = CheckNumInvalid{}
+var _ Step = MarkInvalid{}
+var _ Step = ReportFound{}
+
+// TestBranchRecoveryState walks the BranchRecoveryState through a sequence of
+// steps, verifying that:
+//   - the horizon is properly expanded in response to found addrs
+//   - report found children below or equal to previously found causes no change
+//   - marking invalid children expands the horizon
+func TestBranchRecoveryState(t *testing.T) {
+
+	const recoveryWindow = 10
+
+	recoverySteps := []Step{
```

Comment:
Really dig the usage of interfaces here to create more-or-less a DSL for testing the recovery state here!